Formula/cppcheck.rb

```diff
@@ -60,7 +82,72 @@ def install
         cout << "Hello World!" << endl;
         return 0;
       }
+
+      class Example
+      {
+        public:
+          int GetNumber() const;
+          explicit Example(int initialNumber);
+        private:
+          int number;
+      };
+
+      Example::Example(int initialNumber)
+      {
+        number = initialNumber;
+      }
+    EOS
+    system "#{bin}/cppcheck", test_cpp_file
+
+    # Test the "out of bounds" check
+    test_cpp_file_check = testpath/"testcheck.cpp"
+    test_cpp_file_check.write <<-EOS.undent
+      int main()
+      {
+      char a[10];
+      a[10] = 0;
+      return 0;
+      }
     EOS
-    system "#{bin}/cppcheck", "test.cpp"
+    output = shell_output("#{bin}/cppcheck #{test_cpp_file_check} 2>&1")
+    assert_match "out of bounds", output, "Expected #{name} to detect out of bounds"
+
+    # Test the addon functionality: sampleaddon.py imports the cppcheckdata python
+    # module and uses it to parse a cppcheck dump into an OOP structure. We then
+    # check the correct number of detected tokens and function names.
+    cppcheck_module = "#{name}data"
+    expect_token_count = 55
+    expect_function_names = "main,GetNumber,Example"
+    assert_parse_message = "Error: sampleaddon.py: failed: can't parse the #{name} dump."
+
+    sample_addon_file = testpath/"sampleaddon.py"
+    sample_addon_file.write <<-EOS.undent
+      #!/usr/bin/env python
+      """A simple test addon for #{name}, prints function names and token count"""
+      import sys
+      import imp
+      # Manually import the '#{cppcheck_module}' module
+      CFILE, FNAME, CDATA = imp.find_module("#{cppcheck_module}", ["#{addons_dir}"])
+      CPPCHECKDATA = imp.load_module("#{cppcheck_module}", CFILE, FNAME, CDATA)
+
+      for arg in sys.argv[1:]:
+          # Parse the dump file generated by #{name}
+          configKlass = CPPCHECKDATA.parsedump(arg)
+          if len(configKlass.configurations) == 0:
+              sys.exit("#{assert_parse_message}") # Parse failure
+          fConfig = configKlass.configurations[0]
+          # Pick and join the function names in a string, separated by ','
+          detected_functions = ','.join(fn.name for fn in fConfig.functions)
+          detected_token_count = len(fConfig.tokenlist)
+          # Print the function names on the first line and the token count on the second
+          print "%s\\n%s" %(detected_functions, detected_token_count)
+    EOS
+
+    system "#{bin}/cppcheck", "--dump", test_cpp_file
+    test_cpp_file_dump = test_cpp_file.to_s + ".dump"
+    assert(File.exist?(test_cpp_file_dump), "Expected dump file #{test_cpp_file_dump} \
+to be created by #{name}.")
+    python_addon_output = shell_output "python #{sample_addon_file} #{test_cpp_file_dump}"
+    assert_match /#{expect_function_names}\n#{expect_token_count}/, python_addon_output
```

Comment:
Seems like you don't need a regular expression here. Please use a string, i.e. `"#{expect_function_names}\n#{expect_token_count}"`, to avoid accidental misinterpretation of special characters that might be contained in the interpolated strings in the future (currently, it wouldn't affect the outcome).